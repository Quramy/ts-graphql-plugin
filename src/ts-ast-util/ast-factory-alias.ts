import ts from 'typescript';

export const astf = ts.factory;
