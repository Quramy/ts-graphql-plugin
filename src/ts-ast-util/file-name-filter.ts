import _path from 'node:path';

import { globToRegExp } from '../string-util/glob-to-regexp';

export function createFileNameFilter({
  specs,
  projectName,
  _forceWin32 = false,
}: {
  specs: string[] | undefined;
  projectName: string;
  _forceWin32?: boolean;
}) {
  const path = _forceWin32 ? _path.win32 : _path;

  const testers = (specs ?? []).map(pattern => {
    if (pattern.includes('*') || pattern.includes('?')) {
      const regexp = globToRegExp(pattern);
      return (normalized: string) => regexp.test(normalized);
    } else {
      const dirOrFileName = pattern[pattern.length - 1] === '/' ? pattern.slice(0, pattern.length - 1) : pattern;
      return (normalized: string) => normalized === dirOrFileName || normalized.startsWith(dirOrFileName + '/');
    }
  });

  const projectRootDirName = path.dirname(projectName);

  const match = (fileName: string) => {
    const normalized = path.relative(projectRootDirName, fileName).replace(/\\/g, '/');
    return testers.some(tester => tester(normalized));
  };

  return match;
}
