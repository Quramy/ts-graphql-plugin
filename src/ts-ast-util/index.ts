import ts from 'typescript/lib/tsserverlibrary';

export * from './types';

export * from './template-expression-resolver';

export { createScriptSourceHelper } from './script-source-helper';

export type TagCondition = string;

export function findNode(sourceFile: ts.SourceFile, position: number): ts.Node | undefined {
  function find(node: ts.Node): ts.Node | undefined {
    if (position >= node.getStart() && position < node.getEnd()) {
      return ts.forEachChild(node, find) || node;
    }
  }
  return find(sourceFile);
}

export function findAllNodes(sourceFile: ts.SourceFile, cond: (n: ts.Node) => boolean): ts.Node[] {
  const result: ts.Node[] = [];
  function find(node: ts.Node) {
    if (cond(node)) {
      result.push(node);
      return;
    } else {
      ts.forEachChild(node, find);
    }
  }
  find(sourceFile);
  return result;
}

export function isTagged(node: ts.Node, condition: TagCondition) {
  if (!node || !node.parent) return false;
  if (!ts.isTaggedTemplateExpression(node.parent)) return false;
  const tagNode = node.parent;
  return tagNode.tag.getText() === condition;
}

export function isTemplateLiteralTypeNode(node: ts.Node): node is ts.TemplateLiteralTypeNode {
  // ts.isNoSubstitutionTemplateLiteral exists TypeScript >= 4.1
  return typeof ts.isTemplateLiteralTypeNode === 'function' && ts.isTemplateLiteralTypeNode(node);
}
