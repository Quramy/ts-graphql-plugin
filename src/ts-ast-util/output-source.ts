import path from 'path';
import ts from 'typescript';

import { OutputSource } from './types';
import {
  createImportSpecifier,
  isImportDeclarationWithCondition,
  mergeImportDeclarationsWithSameModules,
} from './utilily-functions';

const printer = ts.createPrinter();

/**
 *
 * @internal
 *
 */
export class DefaultOutputSource implements OutputSource {
  private _unmergedComment: string[] = [];
  private _statements: ts.Statement[] = [];

  constructor(private _filename: string) {}

  get outputFileName() {
    return this._filename;
  }

  get outputDirName() {
    return path.dirname(this._filename);
  }

  findImportDeclarationIndex({ isDefault, name, from }: { isDefault?: boolean; name?: string; from?: string }) {
    return this._statements.findIndex(st => isImportDeclarationWithCondition(st, { isDefault, name, from }));
  }

  pushNamedImportIfNeeded(identifierName: string, from: string) {
    if (this.findImportDeclarationIndex({ name: identifierName, from }) !== -1) return false;
    const importSpecifier = createImportSpecifier(false, undefined, ts.createIdentifier(identifierName));
    const namedBinding = ts.createNamedImports([importSpecifier]);
    const importClause = ts.createImportClause(undefined, namedBinding);
    const moduleSpecifier = ts.createStringLiteral(from);
    const importDeclaration = ts.createImportDeclaration(undefined, undefined, importClause, moduleSpecifier);
    const indexOfSameModuleImport = this.findImportDeclarationIndex({ from });
    if (indexOfSameModuleImport === -1) {
      this.pushImportDeclaration(importDeclaration);
    } else {
      const base = this.getStatements()[indexOfSameModuleImport] as ts.ImportDeclaration;
      const merged = mergeImportDeclarationsWithSameModules(base, importDeclaration);
      this._replaceStatement(indexOfSameModuleImport, merged);
    }
    return true;
  }

  pushDefaultImportIfNeeded(identifierName: string, from: string) {
    if (this.findImportDeclarationIndex({ name: identifierName, from, isDefault: true }) !== -1) return false;
    const importClause = ts.createImportClause(ts.createIdentifier(identifierName), undefined);
    const moduleSpecifier = ts.createStringLiteral(from);
    const importDeclaration = ts.createImportDeclaration(undefined, undefined, importClause, moduleSpecifier);
    const indexOfSameModuleImport = this.findImportDeclarationIndex({ from });
    if (indexOfSameModuleImport === -1) {
      this.pushImportDeclaration(importDeclaration);
    } else {
      const base = this.getStatements()[indexOfSameModuleImport] as ts.ImportDeclaration;
      const merged = mergeImportDeclarationsWithSameModules(base, importDeclaration);
      this._replaceStatement(indexOfSameModuleImport, merged);
    }
    return true;
  }

  pushImportDeclaration(declaration: ts.ImportDeclaration) {
    for (let i = 0; i < this._statements.length; i++) {
      if (!ts.isImportDeclaration(this._statements[i])) {
        this._insertStatement(i, declaration);
        return;
      }
    }
    this._insertStatement(0, declaration);
  }

  pushStatement(statement: ts.Statement) {
    this._insertStatement(this._statements.length, statement);
  }

  replaceStatement(statement: ts.Statement, newStatement: ts.Statement) {
    const idx = this._findIndex(statement);
    if (idx === -1) return false;
    this._replaceStatement(idx, newStatement);
    return true;
  }

  removeStatement(statement: ts.Statement) {
    const idx = this._findIndex(statement);
    if (idx === -1) return false;
    this._statements = [...this._statements.slice(0, idx), ...this._statements.slice(idx + 1)];
    return true;
  }

  writeLeadingComment(comment: string) {
    if (this._statements.length > 0) {
      ts.addSyntheticLeadingComment(this._statements[0], ts.SyntaxKind.MultiLineCommentTrivia, ` ${comment} `, true);
    } else {
      this._unmergedComment.push(comment);
    }
  }

  getStatements() {
    return this._statements as ReadonlyArray<ts.Statement>;
  }

  toSourceFile() {
    const sourceFile = ts.createSourceFile(this._filename, '', ts.ScriptTarget.Latest, false, ts.ScriptKind.TS);
    return ts.updateSourceFileNode(sourceFile, this._statements);
  }

  toFileContent() {
    const content = printer.printFile(this.toSourceFile());
    return { fileName: this._filename, content };
  }

  private _findIndex(statement: ts.Statement) {
    return this._statements.findIndex(st => st === statement);
  }

  private _insertStatement(index: number, statement: ts.Statement) {
    this._statements = [...this._statements.slice(0, index), statement, ...this._statements.slice(index)];
    if (this._unmergedComment.length) {
      this._unmergedComment.forEach(comment => this.writeLeadingComment(comment));
      this._unmergedComment = [];
    }
  }

  private _replaceStatement(index: number, statement: ts.Statement) {
    if (!this._statements[index]) return;
    this._statements[index] = statement;
  }
}

export function createOutputSource({ outputFileName }: { outputFileName: string }) {
  return new DefaultOutputSource(outputFileName);
}
