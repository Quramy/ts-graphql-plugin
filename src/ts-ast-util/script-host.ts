import ts from 'typescript';

export class ScriptHost implements ts.LanguageServiceHost {
  private readonly _fileMap = new Map<string, string | undefined>();
  private readonly _fileVersionMap = new Map<string, number>();

  constructor(private readonly _currentDirectory: string, private readonly _compilerOptions: ts.CompilerOptions) {}

  readFile(fileName: string) {
    const hit = this._fileMap.get(fileName);
    if (hit != null) return hit;
    return this.loadFromFileSystem(fileName);
  }

  loadFromFileSystem(fileName: string) {
    const content = ts.sys.readFile(fileName, 'uts8');
    this._updateFile(fileName, content);
    return content;
  }

  getCurrentDirectory() {
    return this._currentDirectory;
  }

  getScriptSnapshot(fileName: string) {
    const file = this._fileMap.get(fileName);
    if (file == null) return;
    return ts.ScriptSnapshot.fromString(file);
  }

  getScriptVersion(fileName: string) {
    const version = this._fileVersionMap.get(fileName);
    if (!version) return '0';
    return version + '';
  }

  getScriptFileNames() {
    return [...this._fileMap.keys()];
  }

  getCompilationSettings() {
    return this._compilerOptions;
  }

  getDefaultLibFileName(opt: ts.CompilerOptions) {
    return ts.getDefaultLibFileName(opt);
  }

  // Sinse ts 4.7, imeplement required.
  // Delegating to `ts.sys.fileExists(path)` is not for me, but I don't know whether the following implementation is correct.
  fileExists() {
    return false;
  }

  protected _updateFile(fileName: string, content: string | undefined) {
    this._fileMap.set(fileName, content);
    const currentVersion = this._fileVersionMap.get(fileName) || 0;
    this._fileVersionMap.set(fileName, currentVersion + 1);
  }
}
