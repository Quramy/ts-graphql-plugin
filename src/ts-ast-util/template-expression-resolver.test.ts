import * as ts from 'typescript';
import { mark, Frets } from 'fretted-strings';

import { findAllNodes } from '.';
import { TemplateExpressionResolver } from '../ts-ast-util/template-expression-resolver';
import { createTestingLanguageService } from './testing/testing-language-service';

describe(TemplateExpressionResolver.prototype.resolve, () => {
  it('should resolve for empty NoSubstitutionTemplateLiteral node', () => {
    const langService = createTestingLanguageService({
      files: [
        {
          fileName: 'main.ts',
          content: 'const query = ``',
        },
      ],
    });
    const source = langService.getProgram()!.getSourceFile('main.ts');
    if (!source) return fail();
    const [node] = findAllNodes(source, node => ts.isNoSubstitutionTemplateLiteral(node));
    const resolver = new TemplateExpressionResolver(langService, () => '');
    const actual = resolver.resolve('main.ts', node as ts.NoSubstitutionTemplateLiteral).resolvedInfo!;
    if (!actual) return fail();
    expect(actual.combinedText).toBe('');
    expect(() => actual.getInnerPosition(0)).toThrowError();
    expect(() => actual.getInnerPosition(14)).toThrowError();
    expect(actual.getInnerPosition(15)).toStrictEqual({ fileName: 'main.ts', pos: 0 });
    expect(actual.getSourcePosition(0)).toStrictEqual({ fileName: 'main.ts', pos: 15 });
  });

  it('should resolve for no closed NoSubstitutionTemplateLiteral node', () => {
    const frets = {} as Frets;
    const langService = createTestingLanguageService({
      files: [
        {
          fileName: 'main.ts',
          // prettier-ignore
          content: mark(
            'const query = `query { }' + '\n' +
            '%%%           ^        ^   %%%' + '\n' +
            '%%%           a1       a2  %%%',
            frets,
          ),
        },
      ],
    });
    const source = langService.getProgram()!.getSourceFile('main.ts');
    if (!source) return fail();
    const [node] = findAllNodes(source, node => ts.isNoSubstitutionTemplateLiteral(node));
    const resolver = new TemplateExpressionResolver(langService, () => '');
    const actual = resolver.resolve('main.ts', node as ts.NoSubstitutionTemplateLiteral).resolvedInfo!;
    if (!actual) return fail();

    expect(actual.combinedText).toBe(
      // prettier-ignore
      mark(
        'query { }' + '\n' +
        '%%%     ^  %%%' + '\n' +
        '%%%     b2 %%%',
        frets,
      ),
    );

    expect(() => actual.getInnerPosition(frets.a1.pos)).toThrowError();
    expect(actual.getInnerPosition(frets.a1.pos + 1)).toStrictEqual({ fileName: 'main.ts', pos: 0 });
    expect(actual.getSourcePosition(0)).toStrictEqual({ fileName: 'main.ts', pos: frets.a1.pos + 1 });

    expect(actual.getInnerPosition(frets.a2.pos + 1)).toStrictEqual({ fileName: 'main.ts', pos: frets.b2.pos + 1 });
    expect(() => actual.getInnerPosition(frets.a2.pos + 2)).toThrowError();
    expect(actual.getSourcePosition(frets.b2.pos + 1)).toStrictEqual({
      fileName: 'main.ts',
      pos: frets.a2.pos + 1,
    });
    expect(() => actual.getSourcePosition(frets.b2.pos + 2)).toThrowError();
  });

  it('should resolve for closed NoSubstitutionTemplateLiteral node', () => {
    const frets = {} as Frets;
    const langService = createTestingLanguageService({
      files: [
        {
          fileName: 'main.ts',
          // prettier-ignore
          content: mark(
            'const query = `query { }`' + '\n' +
            '%%%           ^        ^   %%%' + '\n' +
            '%%%           a1       a2  %%%',
            frets,
          ),
        },
      ],
    });
    const source = langService.getProgram()!.getSourceFile('main.ts');
    if (!source) return fail();
    const [node] = findAllNodes(source, node => ts.isNoSubstitutionTemplateLiteral(node));
    const resolver = new TemplateExpressionResolver(langService, () => '');
    const actual = resolver.resolve('main.ts', node as ts.NoSubstitutionTemplateLiteral).resolvedInfo!;
    if (!actual) return fail();

    expect(actual.combinedText).toBe(
      // prettier-ignore
      mark(
        'query { }' + '\n' +
        '%%%     ^  %%%' + '\n' +
        '%%%     b2 %%%',
        frets,
      ),
    );

    expect(() => actual.getInnerPosition(frets.a1.pos)).toThrowError();
    expect(actual.getInnerPosition(frets.a1.pos + 1)).toStrictEqual({ fileName: 'main.ts', pos: 0 });
    expect(actual.getSourcePosition(0)).toStrictEqual({ fileName: 'main.ts', pos: frets.a1.pos + 1 });

    expect(actual.getInnerPosition(frets.a2.pos + 1)).toStrictEqual({ fileName: 'main.ts', pos: frets.b2.pos + 1 });
    expect(() => actual.getInnerPosition(frets.a2.pos + 2)).toThrowError();
    expect(actual.getSourcePosition(frets.b2.pos + 1)).toStrictEqual({
      fileName: 'main.ts',
      pos: frets.a2.pos + 1,
    });
    expect(() => actual.getSourcePosition(frets.b2.pos + 2)).toThrowError();
  });

  it('should resolve templateExpression spans in no closed TemplateExpression node', () => {
    const frets = {} as Frets;
    const langService = createTestingLanguageService({
      files: [
        {
          fileName: 'main.ts',
          // prettier-ignore
          content: mark('\n' +
            '    const fragment = gql`fragment Hoge on Foo { name }`;'     + '\n' +
            '    const query = gql`'                                       + '\n' +
            '%%%                  ^                                   %%%' + '\n' +
            '%%%                  a1                                  %%%' + '\n' +
            '      ${fragment}'                                            + '\n' +
            '%%%     ^       ^                                        %%%' + '\n' +
            '%%%     a3      a2                                       %%%' + '\n' +
            '      query MyQuery {'                                        + '\n' +
            '        ...Hoge'                                              + '\n' +
            '      }'                                                      + '\n' +
            '      ${fragment}'                                            + '\n' +
            '%%%     ^       ^                                        %%%' + '\n' +
            '%%%     a5      a4                                       %%%' + '\n' +
            '    ', frets),
        },
      ],
    });
    const source = langService.getProgram()!.getSourceFile('main.ts');
    if (!source) return fail();
    const nodes = findAllNodes(source, node => ts.isTaggedTemplateExpression(node));
    const resolver = new TemplateExpressionResolver(langService, () => '');
    const actual = resolver.resolve('main.ts', nodes[1] as ts.TemplateExpression).resolvedInfo!;
    if (!actual) return fail();

    // prettier-ignore
    const expectedCombinedText = mark('\n' +
      '      fragment Hoge on Foo { name }'     + '\n' +
      '%%%                               ^ %%%' + '\n' +
      '%%%                               b1%%%' + '\n' +
      '      query MyQuery {'                   + '\n' +
      '        ...Hoge'                         + '\n' +
      '      }'                                 + '\n' +
      '      fragment Hoge on Foo { name }'     + '\n' +
      '%%%                               ^ %%%' + '\n' +
      '%%%                               b2%%%' + '\n' +
      '    ', frets);

    expect(actual.combinedText).toBe(expectedCombinedText);

    expect(() => actual.getInnerPosition(frets.a1.pos)).toThrowError();
    expect(actual.getInnerPosition(frets.a1.pos + 1)).toStrictEqual({ fileName: 'main.ts', pos: 0 });
    expect(actual.getSourcePosition(0)).toStrictEqual({ fileName: 'main.ts', pos: frets.a1.pos + 1 });

    expect(() => actual.getInnerPosition(frets.a2.pos)).toThrowError();
    expect(actual.getInnerPosition(frets.a2.pos + 1)).toStrictEqual({ fileName: 'main.ts', pos: frets.b1.pos + 1 });
    expect(actual.getSourcePosition(frets.b1.pos)).toStrictEqual({
      fileName: 'main.ts',
      pos: frets.a3.pos,
      isInOtherExpression: true,
    });
    expect(actual.getSourcePosition(frets.b1.pos + 1)).toStrictEqual({
      fileName: 'main.ts',
      pos: frets.a2.pos + 1,
    });

    expect(() => actual.getInnerPosition(frets.a4.pos)).toThrowError();
    expect(actual.getInnerPosition(frets.a4.pos + 1)).toStrictEqual({ fileName: 'main.ts', pos: frets.b2.pos + 1 });
    expect(actual.getSourcePosition(frets.b2.pos)).toStrictEqual({
      fileName: 'main.ts',
      pos: frets.a5.pos,
      isInOtherExpression: true,
    });
    expect(actual.getSourcePosition(frets.b2.pos + 1)).toStrictEqual({
      fileName: 'main.ts',
      pos: frets.a4.pos + 1,
    });
  });

  it('should resolve templateExpression spans in closed TemplateExpression node', () => {
    const frets = {} as Frets;
    const langService = createTestingLanguageService({
      files: [
        {
          fileName: 'main.ts',
          // prettier-ignore
          content: mark('\n' +
            '    const fragment = gql`fragment Hoge on Foo { name }`;'     + '\n' +
            '    const query = gql`'                                       + '\n' +
            '%%%                  ^                                   %%%' + '\n' +
            '%%%                  a1                                  %%%' + '\n' +
            '      ${fragment}'                                            + '\n' +
            '%%%     ^       ^                                        %%%' + '\n' +
            '%%%     a3      a2                                       %%%' + '\n' +
            '      query MyQuery {'                                        + '\n' +
            '        ...Hoge'                                              + '\n' +
            '      }'                                                      + '\n' +
            '      ${fragment}'                                            + '\n' +
            '%%%     ^       ^                                        %%%' + '\n' +
            '%%%     a5      a4                                       %%%' + '\n' +
            '    `;', frets),
        },
      ],
    });
    const source = langService.getProgram()!.getSourceFile('main.ts');
    if (!source) return fail();
    const nodes = findAllNodes(source, node => ts.isTaggedTemplateExpression(node));
    const resolver = new TemplateExpressionResolver(langService, () => '');
    const actual = resolver.resolve('main.ts', nodes[1] as ts.TemplateExpression).resolvedInfo!;
    if (!actual) return fail();

    // prettier-ignore
    const expectedCombinedText = mark('\n' +
      '      fragment Hoge on Foo { name }'     + '\n' +
      '%%%                               ^ %%%' + '\n' +
      '%%%                               b1%%%' + '\n' +
      '      query MyQuery {'                   + '\n' +
      '        ...Hoge'                         + '\n' +
      '      }'                                 + '\n' +
      '      fragment Hoge on Foo { name }'     + '\n' +
      '%%%                               ^ %%%' + '\n' +
      '%%%                               b2%%%' + '\n' +
      '    ', frets);

    expect(actual.combinedText).toBe(expectedCombinedText);

    expect(() => actual.getInnerPosition(frets.a1.pos)).toThrowError();
    expect(actual.getInnerPosition(frets.a1.pos + 1)).toStrictEqual({ fileName: 'main.ts', pos: 0 });
    expect(actual.getSourcePosition(0)).toStrictEqual({ fileName: 'main.ts', pos: frets.a1.pos + 1 });

    expect(() => actual.getInnerPosition(frets.a2.pos)).toThrowError();
    expect(actual.getInnerPosition(frets.a2.pos + 1)).toStrictEqual({ fileName: 'main.ts', pos: frets.b1.pos + 1 });
    expect(actual.getSourcePosition(frets.b1.pos)).toStrictEqual({
      fileName: 'main.ts',
      pos: frets.a3.pos,
      isInOtherExpression: true,
    });
    expect(actual.getSourcePosition(frets.b1.pos + 1)).toStrictEqual({
      fileName: 'main.ts',
      pos: frets.a2.pos + 1,
    });

    expect(() => actual.getInnerPosition(frets.a4.pos)).toThrowError();
    expect(actual.getInnerPosition(frets.a4.pos + 1)).toStrictEqual({ fileName: 'main.ts', pos: frets.b2.pos + 1 });
    expect(actual.getSourcePosition(frets.b2.pos)).toStrictEqual({
      fileName: 'main.ts',
      pos: frets.a5.pos,
      isInOtherExpression: true,
    });
    expect(actual.getSourcePosition(frets.b2.pos + 1)).toStrictEqual({
      fileName: 'main.ts',
      pos: frets.a4.pos + 1,
    });
  });

  describe('resolveErrors', () => {
    it('should return resolve errors when interpolation is too complex', () => {
      const frets = {} as Frets;
      const langService = createTestingLanguageService({
        files: [
          {
            fileName: 'main.ts',
            // prettier-ignore
            content: mark('\n' +
              '    const fragment = gql`fragment Hoge on Foo { name }`;'     + '\n' +
              '    const query = gql`'                                       + '\n' +
              '      ${fn(fragment)}'                                        + '\n' +
              '%%%     ^           ^                                    %%%' + '\n' +
              '%%%     a1          a2                                   %%%' + '\n' +
              '      query MyQuery {'                                        + '\n' +
              '        ...Hoge'                                              + '\n' +
              '      }'                                                      + '\n' +
              '    ', frets),
          },
        ],
      });
      const source = langService.getProgram()!.getSourceFile('main.ts');
      if (!source) return fail();
      const nodes = findAllNodes(source, node => ts.isTaggedTemplateExpression(node));
      const resolver = new TemplateExpressionResolver(langService, () => '');
      const actual = resolver.resolve('main.ts', nodes[1] as ts.TemplateExpression);
      if (actual.resolvedInfo) return fail();
      expect(actual.resolveErrors).toStrictEqual([
        {
          fileName: 'main.ts',
          start: frets.a1.pos,
          end: frets.a2.pos,
        },
      ]);
    });
  });

  describe('string combinination pattern', () => {
    it('should return combined string in NoSubstitutionTemplateLiteral', () => {
      const langService = createTestingLanguageService({
        files: [
          {
            fileName: 'main.ts',
            content: 'const query = `query { }`',
          },
        ],
      });
      const source = langService.getProgram()!.getSourceFile('main.ts');
      if (!source) return fail();
      const [node] = findAllNodes(source, node => ts.isNoSubstitutionTemplateLiteral(node));
      const resolver = new TemplateExpressionResolver(langService, () => '');
      const actual = resolver.resolve('main.ts', node as ts.NoSubstitutionTemplateLiteral).resolvedInfo!;
      expect(actual!.combinedText).toBe('query { }');
    });

    it('should return combined string in TemplateExpression with StringLiteral', () => {
      const langService = createTestingLanguageService({
        files: [
          {
            fileName: 'main.ts',
            content: "const query = `query { ${'foo'} }`",
          },
        ],
      });
      const source = langService.getProgram()!.getSourceFile('main.ts');
      if (!source) return fail();
      const [node] = findAllNodes(source, node => ts.isTemplateExpression(node));
      const resolver = new TemplateExpressionResolver(langService, () => '');
      const actual = resolver.resolve('main.ts', node as ts.TemplateExpression).resolvedInfo!;
      expect(actual!.combinedText).toMatchSnapshot();
    });

    it('should return combined string with reference to other literal', () => {
      const langService = createTestingLanguageService({
        files: [
          {
            fileName: 'main.ts',
            content: `
              const fragment = \`
                fragment Foo on Hoge {
                  name
                }\`;
              const query = \`
                \${fragment}
                query {
                  ...Foo
                }\`;
            `,
          },
        ],
      });
      const source = langService.getProgram()!.getSourceFile('main.ts');
      if (!source) return fail();
      const [node] = findAllNodes(source, node => ts.isTemplateExpression(node));
      const resolver = new TemplateExpressionResolver(langService, () => '');
      const actual = resolver.resolve('main.ts', node as ts.TemplateExpression).resolvedInfo!;
      expect(actual!.combinedText).toMatchSnapshot();
    });

    it('should return combined string with property access interpolation', () => {
      const langService = createTestingLanguageService({
        files: [
          {
            fileName: 'main.ts',
            content: `
              const fragment = \`
                fragment Foo on Hoge {
                  name
                }\`;
              const obj = { fragment: fragment };
              const query = \`
                \${obj.fragment}
                query {
                  ...Foo
                }\`;
            `,
          },
        ],
      });
      const source = langService.getProgram()!.getSourceFile('main.ts');
      if (!source) return fail();
      const [node] = findAllNodes(source, node => ts.isTemplateExpression(node));
      const resolver = new TemplateExpressionResolver(langService, () => '');
      const actual = resolver.resolve('main.ts', node as ts.TemplateExpression).resolvedInfo!;
      expect(actual!.combinedText).toMatchSnapshot();
    });

    it('should return combined string with shorthand property access interpolation', () => {
      const langService = createTestingLanguageService({
        files: [
          {
            fileName: 'main.ts',
            content: `
              const fragment = \`
                fragment Foo on Hoge {
                  name
                }\`;
              const obj = { fragment };
              const query = \`
                \${obj.fragment}
                query {
                  ...Foo
                }\`;
            `,
          },
        ],
      });
      const source = langService.getProgram()!.getSourceFile('main.ts');
      if (!source) return fail();
      const [node] = findAllNodes(source, node => ts.isTemplateExpression(node));
      const resolver = new TemplateExpressionResolver(langService, () => '');
      const actual = resolver.resolve('main.ts', node as ts.TemplateExpression).resolvedInfo!;
      expect(actual!.combinedText).toMatchSnapshot();
    });

    it('should return combined string with class static property interpolation', () => {
      const langService = createTestingLanguageService({
        files: [
          {
            fileName: 'main.ts',
            content: `
              class MyClass {
                static fragment = \`
                  fragment Foo on Hoge {
                    name
                  }
                \`;
              }
              const query = \`
                \${MyClass.fragment}
                query {
                  ...Foo
                }\`;
            `,
          },
        ],
      });
      const source = langService.getProgram()!.getSourceFile('main.ts');
      if (!source) return fail();
      const [node] = findAllNodes(source, node => ts.isTemplateExpression(node));
      const resolver = new TemplateExpressionResolver(langService, () => '');
      const actual = resolver.resolve('main.ts', node as ts.TemplateExpression).resolvedInfo!;
      expect(actual!.combinedText).toMatchSnapshot();
    });

    it('should return combined string with hopping reference', () => {
      const langService = createTestingLanguageService({
        files: [
          {
            fileName: 'main.ts',
            content: `
              const fragment = \`
                fragment Foo on Hoge {
                  name
                }\`;
              const f2 = fragment;
              const query = \`
                \${f2}
                query {
                  ...Foo
                }\`;
            `,
          },
        ],
      });
      const source = langService.getProgram()!.getSourceFile('main.ts');
      if (!source) return fail();
      const [node] = findAllNodes(source, node => ts.isTemplateExpression(node));
      const resolver = new TemplateExpressionResolver(langService, () => '');
      const actual = resolver.resolve('main.ts', node as ts.TemplateExpression).resolvedInfo!;
      expect(actual!.combinedText).toMatchSnapshot();
    });

    it('should return combined string with reference between multiple files', () => {
      const langService = createTestingLanguageService({
        files: [
          {
            fileName: 'fragment.ts',
            content: `
              export const fragment = \`
                fragment Foo on Hoge {
                  name
                }\`;
            `,
          },
          {
            fileName: 'main.ts',
            content: `
              import { fragment } from './fragment';
              const query = \`
                \${fragment}
                query {
                  ...Foo
                }\`;
            `,
          },
        ],
      });
      const source = langService.getProgram()!.getSourceFile('main.ts');
      if (!source) return fail();
      const [node] = findAllNodes(source, node => ts.isTemplateExpression(node));
      const resolver = new TemplateExpressionResolver(langService, () => '');
      const actual = resolver.resolve('main.ts', node as ts.TemplateExpression).resolvedInfo!;
      expect(actual!.combinedText).toMatchSnapshot();
    });
  });

  describe('cache', () => {
    it('should store result to cache', () => {
      const langService = createTestingLanguageService({
        files: [
          {
            fileName: 'fragment.ts',
            content: `
              export const fragment = \`
                fragment Foo on Hoge {
                  name
                }\`;
            `,
          },
          {
            fileName: 'main.ts',
            content: `
              import { fragment } from './fragment';
              const query = \`
                \${fragment}
                query {
                  ...Foo
                }\`;
            `,
          },
        ],
      });
      const [node] = findAllNodes(langService.getProgram()!.getSourceFile('main.ts')!, node =>
        ts.isTemplateExpression(node),
      );
      const resolver = new TemplateExpressionResolver(langService, () => '');
      const firstResult = resolver.resolve('main.ts', node as ts.TemplateExpression);
      expect(resolver._resultCache.has(node)).toBeTruthy();
      expect(resolver.resolve('main.ts', node as ts.TemplateExpression).resolvedInfo).toBe(firstResult.resolvedInfo);
    });
  });
});

describe(TemplateExpressionResolver.prototype.update, () => {
  it('should return updated info arleady resolved with text replacement', () => {
    const frets = {} as Frets;
    const langService = createTestingLanguageService({
      files: [
        {
          fileName: 'main.ts',
          // prettier-ignore
          content: mark('\n' +
            '    const str = `'       + '\n' +
            '         aaaaaaa'        + '\n' +
            '         aaaaaaa'        + '\n' +
            '%%%      ^     ^    %%%' + '\n' +
            '%%%      a1    a2   %%%' + '\n' +
            '         bbbbbbb'        + '\n' +
            '    `;'                  + '\n' +
            '    ', frets),
        },
      ],
    });
    const source = langService.getProgram()!.getSourceFile('main.ts');
    if (!source) return fail();
    const nodes = findAllNodes(source, node => ts.isNoSubstitutionTemplateLiteral(node));
    const resolver = new TemplateExpressionResolver(langService, () => '');
    const originalInfo = resolver.resolve('main.ts', nodes[0] as ts.NoSubstitutionTemplateLiteral).resolvedInfo!;
    if (!originalInfo) return fail();
    const actual = resolver.update(
      originalInfo,
      {
        start: originalInfo.getInnerPosition(frets.a1.pos).pos,
        end: originalInfo.getInnerPosition(frets.a2.pos + 1).pos,
      },
      'xxxxx',
    );
    const expectedStr = mark(
      `
         aaaaaaa
         xxxxx
     %%% ^   ^    %%%
     %%% b1  b2   %%%
         bbbbbbb
    `,
      frets,
    );
    expect(actual.combinedText).toBe(expectedStr);
    expect(actual.getInnerPosition(frets.a1.pos - 1).pos).toBe(frets.b1.pos - 1);
    expect(actual.getInnerPosition(frets.a1.pos).pos).toBe(frets.b1.pos);
    expect(actual.getInnerPosition(frets.a1.pos + 1).pos).toBe(frets.b1.pos);
    expect(actual.getInnerPosition(frets.a2.pos).pos).toBe(frets.b1.pos);
    expect(actual.getInnerPosition(frets.a2.pos + 1).pos).toBe(frets.b2.pos + 1);
    expect(actual.getSourcePosition(frets.b1.pos - 1).pos).toBe(frets.a1.pos - 1);
    expect(actual.getSourcePosition(frets.b1.pos).pos).toBe(frets.a1.pos);
    expect(actual.getSourcePosition(frets.b1.pos + 1).pos).toBe(frets.a1.pos);
    expect(actual.getSourcePosition(frets.b2.pos).pos).toBe(frets.a1.pos);
    expect(actual.getSourcePosition(frets.b2.pos + 1).pos).toBe(frets.a2.pos + 1);
  });
});
