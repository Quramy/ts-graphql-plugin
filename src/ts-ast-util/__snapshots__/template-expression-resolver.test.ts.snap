// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resolve string combinination pattern should return combined string in TemplateExpression with StringLiteral 1`] = `"query { foo }"`;

exports[`resolve string combinination pattern should return combined string with class static property interpolation 1`] = `
"
                
                  fragment Foo on Hoge {
                    name
                  }
                
                query {
                  ...Foo
                }"
`;

exports[`resolve string combinination pattern should return combined string with hopping reference 1`] = `
"
                
                fragment Foo on Hoge {
                  name
                }
                query {
                  ...Foo
                }"
`;

exports[`resolve string combinination pattern should return combined string with property access interpolation 1`] = `
"
                
                fragment Foo on Hoge {
                  name
                }
                query {
                  ...Foo
                }"
`;

exports[`resolve string combinination pattern should return combined string with reference between multiple files 1`] = `
"
                
                fragment Foo on Hoge {
                  name
                }
                query {
                  ...Foo
                }"
`;

exports[`resolve string combinination pattern should return combined string with reference to other literal 1`] = `
"
                
                fragment Foo on Hoge {
                  name
                }
                query {
                  ...Foo
                }"
`;

exports[`resolve string combinination pattern should return combined string with shorthand property access interpolation 1`] = `
"
                
                fragment Foo on Hoge {
                  name
                }
                query {
                  ...Foo
                }"
`;
